packages/job-worker/src/blueprints/context/adlibActions.ts

```diff
@@ -636,4 +638,21 @@ export class ActionExecutionContext
 	async hackGetMediaObjectDuration(mediaId: string): Promise<number | undefined> {
 		return getMediaObjectDuration(this._context, mediaId, this.studioId)
 	}
+
+	async executeTSRAction(
+		deviceId: string,
+		actionId: string,
+		payload: Record<string, any>
+	): Promise<TSR.ActionExecutionResult> {
+		const device = await this._context.directCollections.PeripheralDevices.findOne({
+			_id: new RegExp(`${deviceId}$`),
```

Comment:
I'm not liking this query. if this was called with a `deviceId` of `"0"`, then it would end up with almost any of the configured devices.
In addition, in an environment where there are playout-gateways for other studios (or ones not attached to a studio), it could pick one of those too.

I don't see a studioId property that you could filter on (that is only on the parent devices), but we do have the parent devices already in memory.
So you could do 
```
const parentIds = this._cache.PeripheralDevices.findAll((d) => d.type === PeripheralDeviceType.PLAYOUT).map(
			(d) => d._id
		)
```
and then add `parentDeviceId: { $in: parentIds },` to the query. That will ensure the selected device belongs to the current studio.

As for improving the regex, it might be possible to determine the full name of it. I'm not certain if this will work reliably enough though. It looks like the parent device is always called `${parentId}PlayoutCoreParent`, and the subdevices are always called `${parentId}Playout${childId}`.
I am not 100% sure this will always work as I have a playout-gateway called `99778B8822A2G448H`, but I think that is because it was from pre-r45 which would sometimes allow a gateway to be called a random id.